.github/workflows/build-and-publish.yml

```diff
@@ -0,0 +1,92 @@
+name: Docker Image CI
+
+on:
+  push:
+    tags:
+     - v**
+  pull_request:
+    branches: [ master ]
+     
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+      packages: write
+      
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+      with:
+        submodules: recursive
+      
+    - name: Log into registry ${{ env.REGISTRY }}
+      if: github.event_name == 'push'
```

Comment:
Just for my understanding, can you explain what this line does? Why do we need to limit login to `push` actions?

Reply:
the `event_name` is `push` when a new tag is pushed (or technically any commit, but we're only looking at tags). For builds triggered from a pull request, the `event_name` is `pull_request` iirc.

This lets us skip publishing-related steps for builds that will never publish artifacts.

Reply:
That makes sense - thanks :)